In [2]:
import os
import sys
import glob
import time
import pandas as pd
import json
import datetime

# Data Paths

In [3]:
data_path = "/DATA/7A/xblock/"
df_path = "/home/aakyildiz/tokens/dataframes/"

#
erc20_transactions_path = "ERC20Transactions/uncompressed"
erc721_transactions_path = "ERC721Transactions/uncompressed"
internal_transactions_path = "InternalTransactions/uncompressed"
normal_transactions_path = "NormalTransactions/uncompressed"

#token information
token_info_path = "TokenInfo/uncompressed/"
erc20_filename = "ERC20TokenInfo.csv"
erc721_filename = "ERC721TokenInfo.csv"

#dataframes
internalTransactions_filename = "internalTransactions.df"
erc20Transactions_filename = "erc20Transactions.df"
erc721Transactions_filename = "erc721Transactions.df"


# Get Token Information and Transactions

##  get erc721 token information

In [4]:
#get erc721 token information
erc721InfoDF = pd.read_csv(data_path + token_info_path + erc721_filename, error_bad_lines=False)
erc721InfoDF = erc721InfoDF[erc721InfoDF['decimal'] != "None"].astype({"decimal": float})
print(erc721InfoDF.shape[0])
erc721InfoDF.head()

7691


address              name  symbol  \
0  0x27b4bc90fbe56f02ef50f2e2f79d7813aa8941a7         gego.dego    GEGO   
1  0x3f9d7b15763f57d7f5015901eaccca27ce996bb2        Polysphere  PLYSPR   
2  0x06012c8cf97bead5deae237070f9587f8e7a266d     CryptoKitties      CK   
3  0xbbe7aeb729e65e3854a30d583cee6a28434eb841        Polysphere    POLY   
4  0x2fb5d7dda4f1f20f974a0fdd547c38674e8d940c  KnightStory Item    MyKI   

     decimal  
0    54084.0  
1        5.0  
2  1996597.0  
3        0.0  
4    13250.0

##  get erc20 token information

In [15]:
#get erc20 token information
#line 154118 in this file is problematic for some reason
erc20InfoDF = pd.read_csv(data_path + token_info_path + erc20_filename, error_bad_lines=False)
print(erc20InfoDF.shape[0])
erc20InfoDF.head()

226250


b'Skipping line 154118: expected 5 fields, saw 7\n'


address                  name symbol  \
0  0x6b0359f95796327475ad4f12ae4e1047c3a67fa3         FinFine Token    FFT   
1  0x12d79c345cac7b050a5ff0797b5a607e254c73f5                Buytex    BUX   
2  0x6bc1f3a1ae56231dbb64d3e82e070857eae86045                xensor    XSR   
3  0x10086399dd8c1e3de736724af52587a2044c9fa2  The Midas Touch Gold   TMTG   
4  0xdac17f958d2ee523a2206206994597c13d831ec7            Tether USD   USDT   

                    totalSupply decimal  
0                  214285500000       3  
1                 9999999999999       4  
2  4800000000000000000000000000      18  
3  9161139666073037500000000000      18  
4             20817977739617189       6

## read all internal transactions

In [4]:
#all_files = glob.glob(data_path + internal_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#internalTransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
internalTransactionsDF = pd.read_feather(df_path + internalTransactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(internalTransactionsDF.shape[0])
internalTransactionsDF.head(10)

NameError: name 'df_path' is not defined

## read all erc20 transactions

In [ ]:
#all_files = glob.glob(data_path + erc20_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc20TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc20TransactionsDF = pd.read_feather(df_path + erc20Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc20TransactionsDF.shape[0])
erc20TransactionsDF.head()

## read all erc721 transactions

In [4]:
#all_files = glob.glob(data_path + erc721_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end="", flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc721TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc721TransactionsDF = pd.read_feather(df_path + erc721Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc721TransactionsDF.shape[0])
erc721TransactionsDF.head()

FileNotFoundError: [Errno 2] No such file or directory: '/home/aakyildiz/tokens/dataframes/erc721Transactions.df'

***

***

***

# Analysis

In [11]:
import numpy as np

## Token Statistics

In [150]:
targetInfoDf = erc721InfoDF
targetDF = erc721TransactionsDF

#get contract addresses
contractAdresses = targetDF[targetDF["tokenAddress"].isin(targetInfo["address"].values)]

#count transactions by tokenAddress 
contractAdresses = contractAdresses.tokenAddress.value_counts()
targetInfoDf['number_of transactions'] = targetInfoDf['address'].map(contractAdresses)
targetInfoDf = targetInfoDf.sort_values(by=['number_of transactions'], ascending=False, na_position='last')
#print(targetInfoDf[targetInfoDf["address"] == token_address])

token_address = "0xD1CEeeeee83F8bCF3BEDad437202b6154E9F5405"
token_symbol = "CHZ"
print(targetInfoDf[targetInfoDf["symbol"] == token_symbol])
print(targetInfoDf[targetInfoDf["address"] == token_address])
targetInfoDf.head(30)

                                         address    name symbol  decimal  \
6989  0xc866ac9ab8d434c8fc9ff5c861245e86ede0c07c  Cheese    CHZ      1.0   

      number_of transactions  
6989                       1  
Empty DataFrame
Columns: [address, name, symbol, decimal, number_of transactions]
Index: []


address                        name  \
18    0x0e3a2a1f2146d86a604adc220b4967a898d7fe07        Gods Unchained Cards   
5472  0x629cdec6acc980ebeebea9e5003bcd44db9fc5ce        Gods Unchained Cards   
2     0x06012c8cf97bead5deae237070f9587f8e7a266d               CryptoKitties   
3663  0x6ebeaf8e8e946f0716e6533a6f2cefc83f60e8ab              Gods Unchained   
5455  0x564cb55c655f727b61d9baf258b547ca04e9e548        Gods Unchained Cards   
176   0xbd13e53255ef917da7557db1b7d2d5c38a2efe24                   DozerDoll   
22    0x8853b05833029e3cf8d3cbb592f9784fa43d2a79                Codex Record   
564   0x8c9b261faef3b3c2e64ab5e58e04615f8c788099          LucidSight-MLB-NFT   
60    0x629a673a8242c2ac4b7b8c5d8735fbeac21a6205                      Sorare   
38    0x1d963688fe2209a98db35c67a041524822cf04ff                  MARBLE-NFT   
276   0x1a94fce7ef36bc90959e206ba569a12afbc91ca1                     Crypton   
82    0xf87e31492faf9a91b02ee0deaad50d51d56d5d4d           Decentraland LAND   
1419  0x7fdcd2a1e52f10c28cb7732f46393e297ecadda1                HyperDragons   
197   0x7e789e2dd1340971de0a9bca35b14ac0939aa330      Crypto stamp Edition 1   
36    0xdceaf1652a131f32a821468dc03a92df0edd86ea    MyCryptoHeroes:Extension   
13    0x273f7f8e6489682df756151f5525576e322d51a3         MyCryptoHeroes:Hero   
0     0x27b4bc90fbe56f02ef50f2e2f79d7813aa8941a7                   gego.dego   
24    0xf915bbfbb6c097dc327e64eec55e9ef4d110d627                     Servant   
49    0x7b00ae36c7485b678fe945c2dd9349eb5baf7b6b              Spheroid SPACE   
4     0x2fb5d7dda4f1f20f974a0fdd547c38674e8d940c            KnightStory Item   
364   0xd2f81cd7a20d60c0d558496c7169a20968389b40                   Etherbots   
10    0x60f80121c31a0d46b5279700f9df786054aa5ee5                     Rarible   
5447  0x55de98f2141a44f804df399468cf7133b7e3e8e5        Gods Unchained Cards   
825   0x8bc67d00253fd60b1afcce88b78820413139f4c6               CryptoFlowers   
5253  0x875eaa41917ced7c1e435272888d23f05098fc97                Baroque Game   
314   0x31af195db332bc9203d758c74df5a5c5e597cdb7           CryptoAssaultUnit   
29    0x98ff4553a3dcbf5e4ff8b5d4cc6fb763b6f20174                MCP Resource   
186   0x5d00d312e171be5342067c09bae883f9bcb2003b            EtheremonMonster   
516   0xbfde6246df72d3ca86419628cac46a9d2b60393c          EtheremonAdventure   
118   0x1276dce965ada590e42d62b3953ddc1ddceb0392  Neon District In-Game Item   

       symbol    decimal  number_of transactions  
18       CARD  6957550.0                 7969847  
5472     CARD  6680552.0                 6680552  
2          CK  1996597.0                 5434740  
3663     GODS   603050.0                  603050  
5455     CARD   580736.0                  580736  
176     Dozer   396189.0                  566636  
22         CR   492854.0                  523779  
564     MLBCB   249173.0                  477187  
60        SOR   159064.0                  410831  
38    MRBLNFT    60452.0                  187361  
276       CTN    87896.0                  168897  
82       LAND    92598.0                  163516  
1419       HD    35123.0                  154783  
197       CS1   150000.0                  152824  
36       MCHE    34445.0                  151279  
13       MCHH    20789.0                  144880  
0        GEGO    54084.0                  134868  
24        SVT    38378.0                  134143  
49      SPACE    93881.0                  109155  
4        MyKI    13250.0                  100547  
364    ETHBOT    18407.0                   91019  
10       RARI    53002.0                   78505  
5447     CARD    75825.0                   75825  
825        CF    24681.0                   66980  
5253      BAR    59551.0                   59551  
314       CAU     4917.0                   59195  
29       MCPR    24005.0                   56566  
186     EMONA    56353.0                   56495  
516     EMOND    18237.0                   52476  
118     NDIGI    45517.0      

Some erc721 tokens have a different decimal and number of transactions. Why is that?

**Different**

CryptoKitties => decimal 1996597, number_of transactions 5434740

MARBLE-NFT => decimal 60452, number of transactions 187361

**Same**

Gods Unchained Cards => decimal 6680552, number_of transactions 6680552

Gods Unchained => decimal 603050, res 603050

Gods Unchained Cards => decimal 580736, res 580736



**Dapps by Type**

Games:
https://sorare.com/
https://cryptodozer.io/
https://godsunchained.com/

Marketplace:
https://rarible.com/

### I believe we need relatively recent data ! (Not sure tho)



## Significance of tokens

In [16]:
erc721InfoDF.head(10)
erc20InfoDF.head(10)

erc721Addresses = erc721InfoDF["address"].values
erc20Addresses = erc20InfoDF["address"].values
tokenAddresses = np.concatenate([erc20Addresses, erc721Addresses])
print(len(tokenAddresses))
tokenAddresses

# token transcation ratio
tokenTransactionRatio = 

233941


array(['0x6b0359f95796327475ad4f12ae4e1047c3a67fa3',
       '0x12d79c345cac7b050a5ff0797b5a607e254c73f5',
       '0x6bc1f3a1ae56231dbb64d3e82e070857eae86045', ...,
       '0xf5c8380f5656f7d82b36dfe5fa2f484839e2fc39',
       '0x6721f071cf1a7d10f5309a896ffa1a724a9a705c',
       '0xf74db21536d2ee9b8428bfee0394586771e13bf0'], dtype=object)